doc/man/dibbler-relay.8
.TH dibbler-relay 8 2004-12-11 GNU Dibbler server

.SH NAME
dibbler-relay \- a portable DHCPv6 relay

.SH DESCRIPTION
.B dibbler-relay
is a portable implementation of the DHCPv6 relay. DHCPv6 relays are
proxies, which allow one server to support links, which server is not
directly connected to. There are ports available for Linux 2.4/2.6 systems
as well as MS Windows XP and 2003. They are freely available under 
.B GNU GPL 
version 2 (or later) license.

.SH SYNOPSIS
.B dibbler-relay
[ run | start | stop | status ]

.SH OPTIONS

.I run
- starts relay in the console. Relay can be closed using ctrl-c.

.I start 
- starts relay in daemon mode.

.I stop
- stops running relay.

.I status
- shows status of the relay.

.SH EXAMPLES

Relay forwards DHCPv6 messages between interfaces. Messages from
client are encapsulated and forwarded as RELAY_FORW messages. Replies
from server are received as RELAY_REPL message. After decapsulation,
they are being sent back to clients. 

It is vital to inform server, where this relayed message was
received. DHCPv6 does this using interface-id option. This identifier
must be unique. Otherwise relays will get confused when they will
receive reply from server. Note that this id does not need to be
alligned with system interface id (ifindex). Think about it as
"ethernet segment identifier" if you are using Ethernet network or as
"bss identifier" if you are using 802.11 network.

Let's assume this case: relay has 2 interfaces: eth0 and
eth1. Clients are located on the eth1 network. Relay should receive
data on that interface using well-known ALL_DHCP_RELAYS_AND_SERVER
multicast address (ff02::1:2). Relay also listens on its global
address 2000::123. Packets received on the eth1 should be forwarded on
the eth0 interface, also using multicast address:

.nf
log-level 8
log-mode short

iface eth0 {
  server multicast yes
}

iface eth1 {
  client multicast yes
  client unicast 2000::123
  interface-id 1000
}
.fi

Here is another exmaple. This time messages should be forwarded from
eth1 and eth3 to the eth0 interface (using multicast) and to the eth2
interface (using server's global address 2000::546). Also clients must
use multicasts (the default approach):

.nf
iface eth0 {
  server multicast yes
}

iface eth2 {
  server unicast 2000::456
}

iface eth1 {
  client multicast yes                    
  interface-id 1000
}

iface eth3 {
  client multicast yes                    
  interface-id 1001
}
.fi

.SH FILES
All files are created in the /var/lib/dibbler directory. During operation,
Dibbler saves various file in that directory. Dibbler relay reads
/etc/dibbler/relay.conf file. Log file is named client.log.

.SH STANDARDS
This implementation aims at conformance to the following standards:

.I RFC 3315 
DHCP for IPv6

.I RFC 3736
Stateless DHCPv6

.SH BUGS
Bugs are tracked with bugzilla, available at
\fIhttp://klub.com.pl/bugzilla/\fP. If you belive you have found a
bug, don't hesitate to report it.

.SH AUTHOR
Dibbler was developed as master thesis on the Technical University of
Gdansk by Tomasz Mrugalski and Marek Senderski. Currently Marek has
not enough free time, so this project is being developed by Tomasz
Mrugalski. Author can be reached at thomson@klub.com.pl.

.SH SEE ALSO
There is dibbler-server(8) and dibbler-client(8) manual page. You are
also advised to take a look at project website located at
\fIhttp://klub.com.pl/dhcpv6/\fP.
As far as authors know, this is the only Windows DHCPv6 stateful
implementation available and the only one with relay support. It is
also one of two freely available under Linux. The other Linux
implementation is available at
\fIhttp://dhcpv6.sourceforge.net\fP,
but it is rather outdated and seems not being actively developed.


CHANGELOG

  Dibbler changelog
 ------------------- 

0.8.1 [not-released-yet]
  - Fixed socket binding problem on server that may affect relay scenarios
  - Fixed segmentation fault when interface-id was not defined on relay interface.
  - Makefiles reworked
  - Srv: Implemented support for DNS Update over UDP (ddns-protocol directive)
  - Srv: Implemented support for DNS Update timeout (ddns-timeout directive)
  - Fixed problem with possible PD crash
  - Linux: radvd.conf generation improved in PD
  - Fix: server/client will no longer load DUID empty file (bug #209)
  - Cli: no longer confused when receiving delegated prefix, but there are no suitable interfaces
  - Srv: server now caches assinged prefixes (bug #217)
  - Cli: support for script execution reimplemented (bugs #207, #216)
  - Makefile system rewritten (using autotools now for better portability)
  - Srv: server now calls external script when assigning or releasing delegated prefix (bug #205)
  - FreeBSD,NetBSD,OpenBSD compilation successful (not tested yet, consider it experimental)
  - Google test framework added for unittests. There are only handful tests implemented now,
    but the number is expected to grow
  - Fix: alignment memory reads, tested on ARMv5 (bug #221)
  - Changed default number of addresses allowed per client from millions to 10
  - Srv: number of currently allocated prefixes is set correctly after restart and database reload
  - Several server and client fixes for DNS Update
  - Cli: Obsolete "no-ia" statement removed. Corrected spelling of "preferred" in config parser.


0.8.0 [2011-05-11]
  - Fixed compilation problems
  - Fix: client, server and relay no longer leave PID file after failed start
  - Fix: error in detecting MAC addresses on Mac
  - Fix: several crashes introduced in 0.8.0RC1 fixed
  - Fix: client no longer gets confused when requesting IA, TA and PD
  - Fix: Server now sends back rapid-commit option
  - Srv: It is now possible to specify how server should handle unknown FQDNs from clients
  - Srv: It is now possible to specify DDNS server address, used during DNS Updates
  - Doc: Fixed bug in example in server man page
  - Srv: answers to RENEW are now constructed properly (single SERVERID is enough)
  - All: Inteface link address is now detected properly on Mac OS X
  - All: Many Mac OS port fixes. It should be operational now. 
  - Linux: client now adds addresses with proper preferred/valid lifetimes
  - Linux: client now properly updates addresses during renew/rebind
  - Cli: Client now properly copies prefixes from ADVERTISE to REQUEST (bug #197)
  - Cli: RELEASE is now sent properly if there are only prefixes, and no addresses (bug #196)
  - Cli: It is now possible to define serveral prefixes in one pd
  - Cli: It is now possible to define specific prefix in pd as a hint
  - Cli: T1,T2 are now properly set for PD
  - Cli: Duplicate IA,TA and PD are no longer created in database, when 
    restarting client with existing leases.
  - Cli: Client no longer silently quits, if adding/updating prefix failed.
  - Linux/Mac OS: daemonizing should close stdin/stdout/stderr (bug #188)
  - Linux: detecting link state change no longer takes 100% CPU
  - All: Vendor spec support improved (new definition format is simpler) (bug #173)
  - Cli: client now waits for longer periods, rather than 1 s
  - Srv: Server now creates separate socket for transmission and not use multicast socket (bug #200)
  - Rel: TRelOptRemoteID compilation fix
  - Srv/Cli: DS-List draft updated to latest, approved version (ds-lite-tunnel-10) (bugs #203, #204)

0.8.0RC1 [2010-09-17]
  - Linux: CONFIRM support added
  - Support for Windows 7 added
  - Server reloads its database after shutdown/start sequence
  - Client with enabled unicast now uses global source addresses
  - Linux: console logging now has color support
  - Srv/Cli: DUID in format 00:01:02:03 is now supported in config files (0x010203 format is still supported)
  - It is now possible to configure IAID
  - Support for Mac OS X added (experimental, of course)
  - Server: DS-Lite tunnel options added
  - Client: Partial support for DS-Lite tunnel option added
  - Experimental support for Remote Autoconf added
  - Fixed problem with not populating AddrMgr correctly
  - Fix: after restart, server no longer refuses to handle adresses after first RELEASE.
  - Fix: Notify scripts are now executed, even when no IA received (bug #193)
  - Code cleanup: Singleton managers implemented
  - Code cleanup: Server REPLY generation refactored
  - Code cleanup: Many obsolete classes removed
  - Code cleanup: Obsolete tunnel endpoint removed
  - Code cleanup: Obsolete VS2003 and VS2005 files removed

 0.7.3 [2009-03-09]
  - Linux: Fix for compilation with libc6, version 2.8
  - New timezone option support added (by Petr Pisar) (bug #185)
  - Linux: Syslog support added (by Petr Pisar) (bug #184)
  - Double timezone removal fixed (by Petr Pisar) (bug #183)
  - Timezone support implemented in Linux systems (by Petr Pisar) (bug #182)
  - Interface-ids other than 4 bytes long are now supported (bug #179)
  - Cli/Srv: Empty FQDN option problem fixed (bug #186 and #187)
  - Srv: Client classification implemented
  - Linux: Pid is now stored in pid_t type (bug #180)
  - Cli: /etc/resolv.conf support improved
  - Linux: Race condition in stop/start fixed (bug #181)

 0.7.2 [2008-08-31]
  - Cli: external scripts may now be executed
  - Cli: addresses are now added properly in autodetect-mode
  - Cli: DNS server is now configured properly under Windows XP/Vista
  - Srv: client's FQDN hints may now be honored
  - Srv: support for extra options added
  - Experimental CONFIRM support added (+extra fixes)
  - Cli: requested options are now sent in RENEW messages
  - Cli: elapsed option is now sent in all messages
  - Cli: client does not get confused when running in autodetect mode and there are no suitable interfaces

 0.7.1 [2008-06-18]

Port-linux/dibbler-relay.cpp
/*
 * Dibbler - a portable DHCPv6
 *
 * authors: Tomasz Mrugalski <thomson@klub.com.pl>
 *
 * released under GNU GPL v2 or later licence
 *
 * $Id: dibbler-relay.cpp,v 1.10 2005-10-11 20:52:47 thomson Exp $
 *
 * $Log: not supported by cvs2svn $
 * Revision 1.9  2005/07/31 14:39:40  thomson
 * Minor changes related to 0.4.1 release.
 *
 * Revision 1.8  2005/07/21 21:40:19  thomson
 * PID checking process is improved.
 *
 * Revision 1.7  2005/04/28 21:20:52  thomson
 * Support for multiple relays added.
 *
 * Revision 1.6  2005/02/03 22:42:25  thomson
 * *** empty log message ***
 *
 * Revision 1.5  2005/02/03 22:06:40  thomson
 * Linux startup/pid checking changed.
 *
 * Revision 1.4  2005/02/03 20:09:11  thomson
 * *** empty log message ***
 *
 * Revision 1.3  2005/01/30 23:12:28  thomson
 * *** empty log message ***
 *
 * Revision 1.2  2005/01/30 22:53:28  thomson
 * *** empty log message ***
 *
 * Revision 1.1  2005/01/11 22:53:35  thomson
 * Relay skeleton implemented.
 *
 */

#include <signal.h>
#include "DHCPRelay.h"
#include "Portable.h"
#include "Logger.h"
#include "daemon.h"

TDHCPRelay * ptr;

void signal_handler(int n) {
    Log(Crit) << "Signal received. Shutting down." << LogEnd;
    ptr->stop();
}

int status() {
    int pid = getServerPID();
    if (pid==-1) {
	cout << "Dibbler server: NOT RUNNING." << endl;
    } else {
	cout << "Dibbler server: RUNNING, pid=" << pid << endl;
    }
    
    pid = getClientPID();
    if (pid==-1) {
	cout << "Dibbler client: NOT RUNNING." << endl;
    } else {
	cout << "Dibbler client: RUNNING, pid=" << pid << endl;
    }

    pid = getRelayPID();
    if (pid==-1) {
	cout << "Dibbler relay : NOT RUNNING." << endl;
    } else {
	cout << "Dibbler relay : RUNNING, pid=" << pid << endl;
    }

    return pid;
}

int run() {
    if (!init(RELPID_FILE, WORKDIR)) {
	return -1;
    }

    TDHCPRelay srv(RELCONF_FILE);
    
    ptr = &srv;
    
    // connect signals
    signal(SIGTERM, signal_handler);
    signal(SIGINT, signal_handler);
    
    srv.run();
    
    die(RELPID_FILE);
    return 0;
}

int help() {
    cout << "Usage:" << endl;
    cout << " dibbler-relay ACTION" << endl
	 << " ACTION = status|start|stop|install|uninstall|run" << endl
	 << " status    - show status and exit" << endl
	 << " start     - start installed service" << endl
	 << " stop      - stop installed service" << endl
	 << " install   - Not available in Linux/Unix systems." << endl
	 << " uninstall - Not available in Linux/Unix systems." << endl
	 << " run       - run in the console" << endl
	 << " help      - displays usage info." << endl;
    return 0;
}

int main(int argc, char * argv[])
{
    char command[256];
    int result=-1;
    int len;

    logStart("(RELAY, Linux port)", "Relay", RELLOG_FILE);

    // parse command line parameters
    if (argc>1) {
	len = strlen(argv[1])+1;
	if (len>255)
	    len = 255;
	strncpy(command,argv[1],len);
    } else {
	memset(command,0,256);
    }

    if (!strncasecmp(command,"start",5) ) {
	result = start(RELPID_FILE, WORKDIR);
    } else
    if (!strncasecmp(command,"run",3) ) {
	result = run();
    } else
    if (!strncasecmp(command,"stop",4)) {
	result = stop(RELPID_FILE);
    } else
    if (!strncasecmp(command,"status",6)) {
	result = status();
    } else
    if (!strncasecmp(command,"help",4)) {
	result = help();
    } else
    if (!strncasecmp(command,"install",7)) {
	cout << "Function not available in Linux/Unix systems." << endl;
	result = 0;
    } else
    if (!strncasecmp(command,"uninstall",9)) {
	cout << "Function not available in Linux/Unix systems." << endl;
	result = 0;
    } else
    {
	help();
    }

    logEnd();

    return result;
}
